contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,317 @@
+
+This is the story of how [PostHog apps](/apps) went from three-day MVP into a service that handles billions of events – and all the lessons we learned along the way.
+
+## Part 1: Building an MVP in three days
+
+PostHog apps are the brainchild of our [first ever hire](/blog/posthog-first-five), Marius Andra. 
+
+Marius wanted a Google Analytics-style world map view for PostHog during our Tuscany offsite hackathon, but he had a problem. The map required location data from IP addresses but, while there were many services to do this, connecting them to event data was difficult.
+
+![Map](../images/blog/how-we-built-an-app-server/map.png)
+
+His solution was to build an app platform and the first ever PostHog app, [GeoIP](/apps/geoip-enrichment), but the project vision immediately expanded to enable more modification of the events pipeline, such as backing up data to S3, syncing GitHub stars, or getting customer feedback.
+
+### Arbitrary app code in Python
+
+The three-day Tuscany hackathon led to the ability to write and run arbitrary Python code on events in our pipeline. The app code got downloaded from GitHub, extracted, and integrated into PostHog. It specifically required Python 3.9, because Python 3.8’s `zipimport` didn’t support the zip archives GitHub provides.
+
+An app, named "plugin" at this point, looked like this:
+
+```python
+# exampleplugin/__init__.py
+from posthog.plugins import PluginBaseClass, PosthogEvent, TeamPlugin
+class ExamplePlugin(PluginBaseClass):   
+    def __init__(self, team_plugin_config: TeamPlugin):
+        super().__init__(team_plugin_config)
+        # other per-team init code
+    def process_event(self, event: PosthogEvent):
+        event.properties["hello"] = "world"
+        return event
+    def process_identify(self, event: PosthogEvent):
+        pass
+```
+
+Other features added included a plugin repository, an interface to configure the apps, and a CLI to preconfigure apps for custom installations. Marius built two sample apps with it: the [currency normalizer](/apps/currency-normalization) (convert event properties to a single currency), and the coveted GeoIP app.
+
+This was a promising start, but it didn’t take long for cracks to appear. 
+
+First, it’s nearly impossible to manage Python dependencies via multiple `requirements.txt` files. There's no way to tell which dependencies your plugin's dependencies will install, without first installing them. There is [nothing like `pip --dry-run`](https://github.com/pypa/pip/issues/53). You need to install the `pip` package and run an `__init__.py` script inside the package to get its dependency tree.
+
+Second, all dependencies were installed together. This made it just a matter of time before a plugin overrode an app dependency (e.g. `django`), and brought everything crashing down. This wasn’t going to work for us, so we needed to rebuild. 
+
+## Part 2: Rebuilding in JavaScript
```

Comment:
```suggestion
## 🛠 Part 2: Rebuilding in JavaScript
```

Reply:
I think adding emojis to each of the major sections will help signpost them a bit more